File: 序号/410.分割数组的最大值.kt
class Solution {
	fun splitArray(nums: IntArray, m: Int): Int {
		// 分成m个子数组 要求每个子数组和的最大值最小
		// 子数组最大值的最小值在nums的最大值（即每个数是一组）；子数组最大值的最大值是nums总和（即不分组）
		// 答案必然在这俩之间，用二分找，查看m个分组能否满足这个mid

		// 找最大值和总和
		var left = Int.MIN_VALUE
		var right = 0
		for (i in nums){
			left = maxOf(i,left)
			right+=i
		}
		while (left<right){
			val mid = left+(right-left)/2
			if (binarySearch(nums, mid, m)) {
				right=mid
			}else left=mid+1
		}
		return right
	}

	fun binarySearch(nums:IntArray,target:Int,m:Int):Boolean{
		var sum = 0
		var group = 0
		for (i in nums){
			sum+=i
			if (sum>target){
				sum=i
				group++
			}
		}
		return group+1 <= m
	}
}

File: 序号/410. 分割数组的最大值.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        # 最大最小 一眼二分
        # nums切割连续子数组，根据mid切，看能分成多少组，超过k就不行，让mid大点
     
        def split(mxSum):
            ''' 每组和最大mxSum，可以切成多少组 '''
            cnt=sm=0
            for x in nums:
                if x > mxSum:
                    return False
                if sm+x>mxSum:
                    cnt+=1
                    sm=x
                else:
                    sm+=x
            return cnt+1<=k
        sm=sum(nums)
        n=len(nums)
        l,r=-1,sm+1
        while l+1 < r:
            mid = (l+r)//2
            if split(mid):
                r=mid
            else: # 满足条件 
                l=mid
        return r